xdsl/dialects/smt.py

```diff
@@ -103,13 +113,134 @@ class XOrOp(VariadicBoolOp):
     name = "smt.xor"
 
 
+def _parse_same_operand_type_variadic_to_bool_op(
+    parser: Parser,
+) -> tuple[Sequence[SSAValue], dict[str, Attribute]]:
+    """
+    Parse a variadic operation with boolean result, with format
+    `%op1, %op2, ..., %opN attr-dict : T` where `T` is the type of all
+    operands.
+    """
+    operand_pos = parser.pos
+    operands = parser.parse_comma_separated_list(
+        parser.Delimiter.NONE, parser.parse_unresolved_operand, "operand list"
+    )
+    attr_dict = parser.parse_optional_attr_dict()
+    parser.parse_punctuation(":")
+    operand_types = parser.parse_type()
+    operands = parser.resolve_operands(
+        operands, (operand_types,) * len(operands), operand_pos
+    )
+    return operands, attr_dict
+
+
+def _print_same_operand_type_variadic_to_bool_op(
+    printer: Printer, operands: Sequence[SSAValue], attr_dict: dict[str, Attribute]
+):
+    """
+    Print a variadic operation with boolean result, with format
+    `%op1, %op2, ..., %opN attr-dict : T` where `T` is the type of all
+    operands.
+    """
+    printer.print(" ")
+    printer.print_list(operands, printer.print_ssa_value)
+    if attr_dict:
+        printer.print_string(" ")
+        printer.print_attr_dict(attr_dict)
+    printer.print(" : ", operands[0].type)
+
+
+@irdl_op_definition
+class DistinctOp(IRDLOperation):
+    """
+    This operation compares the operands and returns true iff all operands are not
+    identical to any of the other operands. The semantics are equivalent to the
+    `distinct` operator defined in the SMT-LIB Standard 2.7 in the Core theory.
+
+    Any SMT sort/type is allowed for the operands and it supports a variadic
+    number of operands, but requires at least two. This is because the `distinct`
+    operator is annotated with `:pairwise` which means that `distinct a b c d` is
+    equivalent to
+
+    ```
+    and (distinct a b) (distinct a c) (distinct a d)
+        (distinct b c) (distinct b d) (distinct c d)
+    ```
+    """
+
+    name = "smt.distinct"
+
+    T: ClassVar = VarConstraint("T", irdl_to_attr_constraint(NonFuncSMTType))
+
+    inputs = var_operand_def(RangeOf(T, length=AtLeast(2)))
+    result = result_def(BoolType())
+
+    traits = traits_def(Pure())
+
+    @classmethod
+    def parse(cls, parser: Parser) -> DistinctOp:
+        operands, attr_dict = _parse_same_operand_type_variadic_to_bool_op(parser)
+        op = DistinctOp(*operands)
+        op.attributes = attr_dict
+        return op
+
+    def print(self, printer: Printer):
+        _print_same_operand_type_variadic_to_bool_op(
+            printer, self.inputs, self.attributes
+        )
```

Comment:
I'm surprised we can't handle this with the assembly format, what breaks there?

Reply:
I think we might have no way to access the type of a single operand. `type($inputs)` will print a comma separated list of all the inputs.

Kind of related but I wondered at some point whether it made sense to have an assembly format directive that directly mentioned a variable constraint. For example, here you could write `$T`.

Reply:
Exactly what Alex said :+1: 
I would be really happy if we could just write `$T`. We could do something like having an explicit marker for type variables, like `%T` for instance? So it's obvious what is compatible with MLIR or not

Reply:
I might cook up a prototype, had a look today and it should be fairly easy

Reply:
Would be really cool!

Reply:
As an update, parsing is easy, but printing is not atm as we don't store the values of the constraint variables anywhere (and don't store the extractors to get them either)